### PR TITLE
Remote VR HITL Evaluation

### PR DESCRIPTION
**Pull request by 0mdc**
## Motivation and Context

This changeset adds the capability of controlling HITL evaluation from a remote device.
Remote control is only supported in the `pick_throw_vr` app state, which was previously named `fetch`.

While this allows for controlling HITL from any remote application, we currently only support the [siro_hitl_unity_client](https://github.com/eundersander/siro_hitl_unity_client) Unity application.

See `examples/siro_sandbox/VR_HITL.md` for complete set-up instructions.

Example video _(robot policy not included)_:

https://github.com/facebookresearch/habitat-lab/assets/110583667/6b5d3eda-eec7-4d0e-97a0-db3cee5f46fd

## Features

The following features are added to the HITL tool:
* Server and remote control.
    * VR pose tracking.
    * Gfx-replay streaming.
    * Server-driver drawing of text and shapes.
    * Sending the navmesh to the client (for constraining teleportable locations).
* Grabbing and throwing objects.
    * Mouse/keyboard support.
    * VR support.
    * Reaching motion and throw animation.
* VR-driven navigation.
* Changing human model at run-time.
* Dataset processing tools for Unity (see files in `scripts/unity_dataset_processing`).

## Notes for reviewers

* The changes were ported from [this Habitat 2.5 branch](https://github.com/facebookresearch/habitat-lab/tree/client-server-hitl) (P0). I suggest comparing with this branch for ease of review.
* Because the fetch state machine is not compatible with Habitat 3.0, and because we don't need it, this functionality was removed.
    * The Spot robot uses the policy from `social_rearrange/pop_play.yaml` - it does not intelligently react to the human's actions.
* Spot robot animations were not ported over.

## How Has This Been Tested

* Tested all app states.
* Tested `pick_throw_vr` app state in local mode.
* Tested `pick_throw_vr` app state, in server mode, with mouse and keyboard.
* Tested `pick_throw_vr` app state, in server mode, using a remote VR headset running [this Unity app](https://github.com/eundersander/siro_hitl_unity_client).
* Installed the stack from scratch, following the documentation.

## Types of changes

- **\[Development\]**

## Checklist

- [x] My code follows the code style of this project.
- [x] I have updated the documentation if required.
- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes if required.

### CODE REVIEW

File: examples/siro_sandbox/server/keyframe_utils.py
Comment (0mdc) [inline examples/siro_sandbox/server/keyframe_utils.py:1]: We may want to move this functionality to `habitat-sim` because it is directly coupled with the keyframe implementation.

File: habitat-lab/habitat/articulated_agent_controllers/humanoid_rearrange_controller.py
Comment (0mdc) [inline habitat-lab/habitat/articulated_agent_controllers/humanoid_rearrange_controller.py:268]: This new `calculate_walk_pose_directional` is the `calculate_walk_pose` function that was used in P0.
It ensures that the human avatar follows the transforms coming from the VR headset (which always slightly moves around, and has a forward gaze constraint).

I created a new function to avoid changing how the humanoids move during training or in the other applications.

@xavierpuigf FYI - I'll let you decide what to do with this code moving forward.

File: habitat-lab/habitat/articulated_agent_controllers/humanoid_rearrange_controller.py
Comment (0mdc) [inline habitat-lab/habitat/articulated_agent_controllers/humanoid_rearrange_controller.py:620]: Here, we ignore updating hand poses if we select an animation file that does not contain hand data (such as the default one).

**Review by 0mdc — COMMENTED** [review]
I left some general comments for discussion.

File: examples/siro_sandbox/README_VR.md
Comment (eundersander) [inline examples/siro_sandbox/README_VR.md:1]: I'll review documentation later when it's finalized.

File: examples/siro_sandbox/app_states/app_state_fetch.py
Comment (eundersander) [inline examples/siro_sandbox/app_states/app_state_fetch.py:25]: Can you explain this file at a high level? I was guessing it would resemble the latest AppStateFetch from the P0 demo branch, but when I diff against that file, I see this is missing lots of functionality related to Spot.

If this is some kind of reduced-functionality version of AppStateFetch, then I would suggest we call it something else instead of `app_state_fetch.py`.

File: examples/siro_sandbox/app_states/app_state_fetch.py
Comment (eundersander) [inline examples/siro_sandbox/app_states/app_state_fetch.py:38]: How about a class-level comment/documentation here?

File: examples/siro_sandbox/app_states/app_state_fetch.py
Comment (eundersander) [reply examples/siro_sandbox/app_states/app_state_fetch.py:25]: Related note to reviewers: I would not focus on the differences in this file compared to `main`'s `app_state_fetch.py`. As I understand, Mikael authored this starting with [this file](https://github.com/facebookresearch/habitat-lab/blob/eundersander/SIRo_hitl_Sep5_demo_refactored/examples/siro_sandbox/app_states/app_state_fetch.py) and removing Spot functionality.

File: examples/siro_sandbox/app_states/app_state_fetch.py
Comment (eundersander) [reply examples/siro_sandbox/app_states/app_state_fetch.py:635]: I don't know what is in `hitl_vr_sample_episodes.json.gz`, but in general this kind of application-specific code in an AppState doesn't bother me. AppStateFetch (or whatever we end up renaming this AppState to) shouldn't aim to be generic; nobody should inherit from it or otherwise build on top of it. It is example code for others to customize in-place or copy-paste from.

File: examples/siro_sandbox/controllers/controller_abc.py
Comment (eundersander) [inline examples/siro_sandbox/controllers/controller_abc.py:122]: Thanks for all the spelling fixes!

File: examples/siro_sandbox/controllers/gui_controller.py
Comment (eundersander) [inline examples/siro_sandbox/controllers/gui_controller.py:197]: A class-level docstring for `GuiHumanoidController` would be good. It's okay for it to be really specific, like "a humanoid controller that supports walking, grasping, and throwing objects". That description is probably somewhat wrong and also too vague. If there are keyboard controls built into this class, mention that, too. No need to actually clean up this code right now.

File: examples/siro_sandbox/controllers/gui_controller.py
Comment (eundersander) [inline examples/siro_sandbox/controllers/gui_controller.py:217]: Compared to [this version](https://github.com/facebookresearch/habitat-lab/blob/eundersander/SIRo_hitl_Sep5_demo_refactored/examples/siro_sandbox/controllers/gui_controller.py#L232), you've removed a lot of code that runs in on_environment_reset. Is that safe?

File: examples/siro_sandbox/controllers/gui_controller.py
Comment (eundersander) [reply examples/siro_sandbox/controllers/gui_controller.py:498]: If you're feeling ambitious, an easy way to (somewhat) clean this up is to move the hack up into application-level code (in an AppState). E.g. this class accepts a `navmesh_to_floor_y_fixup` as an init parameter.

At a high level, it's always better to have hacks at application-level code (where users can see and fix them) versus in library-level code (I think of this class as library-level code).

Comment (eundersander) [reply examples/siro_sandbox/controllers/gui_controller.py:498]: Note on your comment about source data: this -0.17 value corresponds to the height of the navmesh above the ground/floor surface, which is related to how the navmesh was built. I don't remember if our navmeshes are currently generated at load time or loaded from a pre-generated navmesh file.

File: examples/siro_sandbox/gui_avatar_switch_helper.py
Comment (eundersander) [inline examples/siro_sandbox/gui_avatar_switch_helper.py:63]: class-level docstring

**Review by eundersander — COMMENTED** [review]
I started reviewing this, but it occurs to me that I'm mostly interested in your changes as compared to the latest version of these files [here](https://github.com/facebookresearch/habitat-lab/tree/eundersander/SIRo_hitl_Sep5_demo_refactored). It's not really practical or meaningful to review all these changes compared to `main`.

File: examples/siro_sandbox/app_states/app_state_fetch.py
Comment (0mdc) [reply examples/siro_sandbox/app_states/app_state_fetch.py:635]: `hitl_vr_sample_episodes.json.gz` was the set of episodes that we used for the P0 project (with the 5 scenes).

File: examples/siro_sandbox/controllers/gui_controller.py
Comment (0mdc) [reply examples/siro_sandbox/controllers/gui_controller.py:217]: Good catch. This specific code was not ported over.

File: examples/siro_sandbox/app_states/app_state_fetch.py
Comment (0mdc) [reply examples/siro_sandbox/app_states/app_state_fetch.py:25]: This is the `AppStateFetch` application from the P0 project. Because the Fetch state machine is not compatible with Habitat 3.0, and because we don't need it, this functionality was removed.

The result is that the Spot robot uses the policy from `social_rearrange/pop_play.yaml` - it does not intelligently react to the human's actions. However, the new features that were added in P0 are working (remote control, VR, grabbing, throwing, teleporting to navmesh, etc.).

Here's what changed:
* All references and code related to `fetch_baselines_controller` was removed (`FetchState,  `FetchBaselinesController`)
* Hard-coded initial agent positions and spawning logic were removed. Additionally, there was code to make sure that the robot and human spawn near each other.
* The resulting dead functions were removed:
    * _make_agent_face_target()
    * _make_robot_near_human()
    * _fetch_object_remote()
    * _get_target_object_bounding_box()
    * _draw_box_in_pos()
    * _get_state_machine_agent_translation()
    * _get_spot_status_text()
    * _viz_fetcher()
    * _update_place_goal()
    * _update_fetcher()

>If this is some kind of reduced-functionality version of AppStateFetch, then I would suggest we call it something else instead of app_state_fetch.py.

Agreed. "Fetch" is a thing of the past at this point.

File: examples/siro_sandbox/README.md
Comment (xavierpuigf) [inline examples/siro_sandbox/README.md]:
```suggestion
Use `--app-state rearrange` to run the rearrange task in the sandbox app. Supported app states: `rearrange` (default), `fetch`, `socialnav`, `free_camera`. These correspond to the different use cases for the HITL tool (see corresponding implementations in `app_states/app_state_<state name>.py` files). Note, `free_camera` is special: it's not a "task" state, it's a state that lets the user control the camera instead of controlling an agent (useful for debugging and policies' behaviour visualization). Also, `tutorial` is a special state that shows a tutorial sequence at the start of every episode to introduce the user to the scene and goals in a human-in-the-loop context. Tutorial should be followed by the task app state (`rearrange`, `fetch` or `socialnav`) and is currently supported only for the rearrange task.
```

File: examples/siro_sandbox/app_states/app_state_fetch.py
Comment (xavierpuigf) [reply examples/siro_sandbox/app_states/app_state_fetch.py]: Why is the state machine not compatible with Habitat 3.0? I believe that the `pop_play.yaml` policy won't be compatible with a pick-throw demo because that high level policy expects a limited set of objects and placing locations. I understand that the previous code was very specific to P0 but being able to use a state machine or something more flexible than a policy seems valuable.

File: examples/siro_sandbox/app_states/app_state_fetch.py
Comment (xavierpuigf) [inline examples/siro_sandbox/app_states/app_state_fetch.py]: For this variable and the one below, can we add a comment of what they mean?

File: examples/siro_sandbox/app_states/app_state_fetch.py
Comment (xavierpuigf) [inline examples/siro_sandbox/app_states/app_state_fetch.py]: Can we add this as a constant or variable in the init function

File: examples/siro_sandbox/controllers/gui_controller.py
Comment (xavierpuigf) [reply examples/siro_sandbox/controllers/gui_controller.py:218]: Agreed! Making suggestions here, since it seems I cannot create a suggested change for all the lines:
```
self._hint_throw_vel_vec = None # The velocity vector at which the object will be thrown 
self._hint_reach_pos = None # The location at which to pick the object. If None, no object should be picked
self._is_picking_reach_pos = None # Same as _hint_reach_pos but is set to None as soon as the humanoid reaches the object
self._iter_hand_pose = 0 # Iterator to compute the hand position when reaching objects. 
```

File: examples/siro_sandbox/controllers/gui_controller.py
Comment (xavierpuigf) [inline examples/siro_sandbox/controllers/gui_controller.py:388]: Note that this line makes lines 378-398 useless. This is used to create an animation when reaching towards an object, and was turned off so that the hand was responsive when using VR. I am fine with just removing all this.

File: examples/siro_sandbox/server/keyframe_utils.py
Comment (xavierpuigf) [inline examples/siro_sandbox/server/keyframe_utils.py:48]: Shall we remove this?

Comment (0mdc) [reply examples/siro_sandbox/server/keyframe_utils.py:48]: I don't feel strongly about these commented blocks.

They will eventually be uncommented to allow skinned meshes to be transmitted to remote clients.

File: examples/siro_sandbox/VR_HITL.md
Comment (eundersander) [inline examples/siro_sandbox/VR_HITL.md:42]: We should just link to `https://github.com/facebookresearch/habitat-lab/blob/hitl-vr/examples/siro_sandbox/README.md#installation` for how to install the HITL app. That should cover installing habitat-lab and habitat-sim dependencies and most of the required data.

**Review by eundersander — COMMENTED** [review]
I'm in the process of testing the install docs. A few initial thoughts so far:

File: examples/siro_sandbox/VR_HITL.md
Comment (0mdc) [reply examples/siro_sandbox/VR_HITL.md:35]: Much better.

File: examples/siro_sandbox/VR_HITL.md
Comment (0mdc) [reply examples/siro_sandbox/VR_HITL.md:49]: I agree. I'll update the documentation in the other repository, and specify to follow these instructions in `VR_HITL.md`.

File: examples/siro_sandbox/README.md
Comment (eundersander) [inline examples/siro_sandbox/README.md:46]: In installation, we should clarify if we require both `habitat-lab` and `habitat-baselines` packages. You might also try to explain how it's not strictly necessary for the tool running at `examples/hitl_tool/hitl_tool.py` to import from the version of habitat-lab/baselines that lives at the parent directory (it just needs to be importable in your current environment), although I'm not sure how to cleanly explain this.

File: examples/siro_sandbox/README.md
Comment (eundersander) [inline examples/siro_sandbox/README.md:66]: I would move this above `Command-line Options` even though you are technically documenting a command-line option. It's really important for folks to fundamentally understand what we mean by use cases. I would rewrite this section a bit to emphasize what we mean by use case (essentially a mini-application; there should be no dependency between use cases, e.g. `app_state_rearrange.py` and `app_state_throw_fetch_vr.py` are isolated from each other).

This is probably a good spot to repeat our advice for users: if you're looking to customize the HITL tool for your own use case (e.g. custom data-collection), a good starting point is to create your own AppState class by copy-pasting from an existing one.

It's unfortunate/confusing that app states and use cases aren't 1-to-1, but I don't want to change that. You already do a good job of explaining the tutorial exception.

File: examples/siro_sandbox/README.md
Comment (eundersander) [inline examples/siro_sandbox/README.md:66]: I'm not sure it's necessary to introduce the term "task state". I don't think of `free_camera` as particularly different from `rearrange` or the other use-cases.

File: examples/siro_sandbox/README.md
Comment (eundersander) [inline examples/siro_sandbox/README.md:67]: We should document the state of the 4 (?) use cases here, near the top of this file. If rearrange and socialnav are works in progress that don't currently run, that's fine, just say so. For `free_camera`, if it works, add a section for it, including keyboard controls and maybe some details like how to specify the episode dataset to load (is there a free_camera.yml config?).

For `pick_throw_vr`, just link them to your separate readme exactly as you already do (but include this here near the top of the file, not buried at the bottom).

**Review by eundersander — COMMENTED** [review]
Some comments on README.md

File: examples/siro_sandbox/README.md
Comment (0mdc) [reply examples/siro_sandbox/README.md:44]: I'll do this in a follow-up PR so that it doesn't get squashed with the rest.

**Review by eundersander — APPROVED** [review]
A few minor snags during install/testing, e.g. hab-sim version, missing websockets, typo in unity_dataset_processing.py command. These should be quick fixes to your readme before merging this.

A major snag is making magnum tools available on Mac. For now, I guess you can recommend Linux and also point folks to the experimental homebrew instructions (see my comment below).

For the rest of my feedback, if you don't have time to address in the near term, I would ask that you collect all the todos into Asana or a doc and post a link here before merging. It can be as simple as a single Asana item, "Remote VR HITL: follow up on PR todos", that links back to this PR with all our unresolved comments.

If all that sounds reasonable to you, then I'm okay giving this the green check. Thanks for doing all this! Hopefully it will help some folks in the new year.

File: examples/siro_sandbox/README.md
Comment (0mdc) [reply examples/siro_sandbox/README.md:71]: `habitat.simulator.habitat_sim_v0.allow_sliding=False` is responsible for 1. I removed the command line argument from the documentation.

**Pull request merged**
